kernel/src/fs/epoll/epoll_file.rs

```diff
@@ -460,3 +448,43 @@ impl Observer<IoEvents> for EpollEntry {
         }
     }
 }
+
+impl Drop for EpollEntry {
+    fn drop(&mut self) {
+        let Some(file) = self.file() else {
+            return;
+        };
+        file.unregister_observer(&(self.self_weak() as _)).unwrap();
+    }
+}
+
+struct KeyableEntry(pub Arc<EpollEntry>);
```

Comment:
> I suggest not using `KeyableEntry` and just using `Arc<EpollEntry>` as key.

I would like to use `Arc<EpollEntry>`, but unfortunately I can't implement `Borrow<_>` for `Arc<EpollEntry>`. I can implement it for `EpollEntry`, but it does not work because what I want to use is `BTreeSet<Arc<EpollEntry>>` instead of `BTreeSet<EpollEntry>`.

> The name `KeyableEntry` makes me think it wraps `KeyableArc`.

I also wanted to give it a better name. However, I don't have an idea at the moment.

> In your case, both `PartialOrd` and `Eq` lend themselves well to being derived.

I'm afraid this will just give wrong implementations, since as documented [here](https://doc.rust-lang.org/beta/std/cmp/trait.PartialOrd.html#derivable):
>> When derived on structs, it will produce a [lexicographic](https://en.wikipedia.org/wiki/Lexicographic_order) ordering based on the top-to-bottom declaration order of the struct’s members.

Do you know how to use `Derive` without triggering the above behavior?

#### Alternative

Alternatively, if you don't like the current implementation using `BTreeSet`, I can replace it with something like `BTreeMap<(FileDesc, KeyableWeak<dyn FileLike>), Arc<EpollEntry>>` (well, the type name is too long, that's unfortunate).

I don't think either way is a clear winner, though I personally have a preference for the former over the latter because it avoids unnecessary memory footprint (we don't need to store the key twice), but it does introduce some concerns like the ones mentioned here that I don't know how to address.

Reply:
Thanks for your explanantion!

> Do you know how to use Derive without triggering the above behavior?

I used to think deriving `Ord` and `PartialOrd` will share the same code, but it seems actually they are implemented in an unrelated way. So manually implementing `Ord` and `PartialOrd` is unavoidable. 

> I don't think either way is a clear winner, though I personally have a preference for the former over the latter because it avoids unnecessary memory footprint (we don't need to store the key twice), but it does introduce some concerns like the ones mentioned here that I don't know how to address.

Yes, the former looks better for me too.